### ISSUE TITLE
Mesh shader discussion

### ISSUE THREAD

**Issue by Venemo** [zeux/niagara#30]
Not sure what is the best place to talk about this so decided maybe we can discuss it here. Hope this is okay.

Looking at the current code, I noticed that the mesh shader workgroup size is 64, but the shader has: `max_vertices = 64, max_primitives = 124` this means that the shader is going to have poor occupancy on AMD HW, effectively leaving 50% of shader invocations under-utilized. Note that this is also suboptimal on NVidia HW which prefers a workgroup size of 32.

I recommend to have a compile-time constant for each of these values (similar to what you do for `MESH_WGSIZE`) and configure it like this:

* AMD: workgroup size == max output vertices == max output primitives >= 64, all outputs accessed by local invocation index
* Nvidia: workgroup size = 32, max output vertices = 64, max output primitives >= 64

You can achieve this by using a "compile-time loop" (a loop using the compile-time constants) which will be optimal on both AMD and NVidia.

**Comment by Venemo** [comment]
Also, you don't have to hardcode these in your program, you can use `maxPreferredMeshWorkGroupInvocations` and `prefersLocalInvocation*Output` from the device properties to determine these things.

**Comment by zeux** [comment]
One issue with this is that all of this requires rebuilding meshlet data, something that would ideally be done offline. It looks as if on AMD hardware specifically the shader is export-bound atm, I've looked a little bit at adding per-triangle culling and it does help performance significantly.

Without per-triangle culling I don't seem to get any benefit from moving to the same number of primitives; with it however I do get better throughput with 64 max primitives, but that's a little too low. I'll test different configurations when I get time. Thanks for the suggestion!

**Comment by zeux** [comment]
I'm also wondering what happens on NV specifically from work groups with size 64 vs 32 - and whether the shader that the driver runs is substantially different wrt performance from a shader that uses work group size 32 but has to process two vertices per invocation (which is what the shader I used for NV extension did, but there it wasn't possible to test wider groups because NV extension requires work group size of 32 if I'm not mistaken).

**Comment by Venemo** [comment]
To your NVidia question: this is explained in one of their mesh shader blogs. As far as I understand NVidia's problem is that it doesn't have proper workgroups, so the whole mesh shader workgroup is executed in a single warp and they emulate the workgroup using a loop.

Therefore, you can get closer to what NVidia hardware actually runs if you use a workgroup size that matches their warp size.

**Comment by zeux** [comment]
> No, it doesn't require rebuilding meshlet data. A workable compromise is if you use a meshlet size of 64 (max 64 vertices and max 64 primitives). In this case on NVidia you would output 1 meshlet per workgroup and on AMD you could output 2 meshlets per workgroup.

What I meant is that varying the max sizes between vendor requires different meshlet data. Just varying workgroup configurations doesn't of course.

64 & 64 is a little problematic depending on the mesh topology - I'd expect that the 64 vertex limit leads to an effective primitive count between 64 and 98 (98 corresponds to an 8x8 grid). Setting the primitive count limit of 64 limits to something like 45 vertices per meshlet for smooth meshes, so you end up underutilizing the threads for vertex transformation.

One other alternative is something like 128 vertices and 192 primitives, which is more balanced wrt the ratio, but still problematic because now this means we need to write all vertex data to LDS :)

> To your NVidia question: this is explained in one of their mesh shader blogs. As far as I understand NVidia's problem is that it doesn't have proper workgroups, so the whole mesh shader workgroup is executed in a single warp and they emulate the workgroup using a loop.

Right, but a work group of 64 would be compiled into two sequential invocations of 32 elements each, vs a shader that uses more or less the same loop if it needs to process a meshlet with >32 vertices/primitives. I understand that using a work group of 64 doesn't match the hardware perfectly, but the question is where the resulting inefficiencies come from.

**Comment by Venemo** [comment]
> What I meant is that varying the max sizes between vendor requires different meshlet data. Just varying workgroup configurations doesn't of course.

Yes, the trick is to find a meshlet size which can work fine on both vendors, and then you can use the same meshlet size but with a slightly different workgroup config.

> 64 & 64 is a little problematic depending on the mesh topology - I'd expect that the 64 vertex limit leads to an effective primitive count between 64 and 98 (98 corresponds to an 8x8 grid). Setting the primitive count limit of 64 limits to something like 45 vertices per meshlet for smooth meshes, so you end up underutilizing the threads for vertex transformation.
> 
> One other alternative is something like 128 vertices and 192 primitives, which is more balanced wrt the ratio, but still problematic because now this means we need to write all vertex data to LDS :)

I think it's worth to experiment with a meshlet size of: max vertices = 128, max primitives = 128 and then use a 128-sized workgroup on AMD and 32 (or 64?) on NVidia.

> Right, but a work group of 64 would be compiled into two sequential invocations of 32 elements each, vs a shader that uses more or less the same loop if it needs to process a meshlet with >32 vertices/primitives. I understand that using a work group of 64 doesn't match the hardware perfectly, but the question is where the resulting inefficiencies come from.

Unfortunately I don't know any more details other than what I told above, only that this is their recommendation.

**Comment by Venemo** [comment]
> I think it's worth to experiment with a meshlet size of: max vertices = 128, max primitives = 128 and then use a 128-sized workgroup on AMD and 32 (or 64?) on NVidia.

One more thought about this. If you definitely don't want to increase the number of max output vertices but you want to use max 128 output primitives, it is still worth it (on AMD) to increase the workgroup size to 128 and make your primitive processing more parallel than it currently is.

**Comment by zeux** [comment]
Can you elaborate on why on AMD there's a benefit to going above 64? It's not intuitively obvious that this should help as 64 (and sometimes 32) is the HW wavefront size.

**Comment by Venemo** [comment]
On RDNA2 each invocation can only really create max 1 vertex and 1 primitive. Any other kind of access pattern is emulated by the driver. This also implies that it may need to launch more invocations than your specified workgroup size in order to fit a larger output.

If you have a workgroup size of 64 but a max primitive count of 126 then the "real" workgroup size will be 126 (this fits 2 waves, which have 128 invocations):

* First 64 invocations will execute the code you wrote
* Next 62 invocations will sit there, just waiting for the first 64 to finish executing your code, then they will output 1 primitive
* The last 2 invocations will be just sitting there deactivated the whole time

So, in fact there are 128 invocations running but you don't utilize all of them. It is more efficient to write your code in a manner that utilizes all invocations instead of letting them sit there doing nothing most of the time.

I try to explain this in my blog post "How mesh shaders are implemented in an AMD driver".

**Comment by Venemo** [comment]
It seems that a few others also struggle to understand this, eg. GravityMark has the same problem. So I think I explained it poorly... Can you suggest a good way to edit my blog post to clarify this?

**Comment by zeux** [comment]
By the way, at least in radv it looks like mesh shaders are always compiled with wave size 64. Do you know if this is a hardware restriction or a driver limitation? I can't currently test any other AMD drivers with mesh shading support...

The reason I ask is I was hoping for something like max_vertices=64 max_triangles=96 to work reasonably well with wave32 but it looks like this is inefficient as it effectively uses the same wave configuration as max_vertices=64 max_triangles=124.

**Comment by zeux** [comment]
Also based on https://github.com/GPUOpen-Drivers/llpc/commit/772eef3ecbb5d294ba033a1da00a06526a3a31e1 my understanding is that on GFX11 (RDNA3) row export would allow emitting more than one vertex or primitive per thread, which would be great as it would provide the much needed flexibility wrt performance.

**Comment by Venemo** [comment]
> By the way, at least in radv it looks like mesh shaders are always compiled with wave size 64. Do you know if this is a hardware restriction or a driver limitation?

It's just the default in our driver. You can use the `RADV_PERFTEST=gewave32` environment variable to use Wave32 mode for geometry processing shaders.

> The reason I ask is I was hoping for something like max_vertices=64 max_triangles=96 to work reasonably well with wave32

Worth a try. Yes it would be inefficient in Wave64 mode. Maybe we should add special casing for 32 and 96.

> my understanding is that on GFX11 (RDNA3) row export would allow emitting more than one vertex or primitive per thread

This is correct, but I haven't implemented that in RADV yet. (I am on vacation this week and will get back to work next week.) However, it will still need some shuffling between SIMD lanes.

> Not sure if GFX11 has other relevant changes for mesh shading.

Yes, it also has a new "fast launch" mode, which will eliminate the need for launching shader invocations that "do nothing".

**Comment by zeux** [comment]
I've switched to an RDNA3 GPU, didn't get a chance to look into this a lot yet but a couple observations:

- It looks like the performance is much more consistent between different meshlet configurations; this should make it much easier to tune which is good.
- Software triangle culling in mesh shader doesn't seem as relevant as it was on RDNA2. This is also great, because it reduces the amount of tradeoffs involved wrt occupancy/ALU et al.

Overall things look closer to what I expected from an NV GPU so that's good news.

There's also two issues I ran into:

- On radv, it looks like the mesh shading path in the default configuration (64/64) is a little slower vs AMDVLK with the same code. I see something like ~7.9ms for the render pass (4K full screen, 7900 GRE) with radv, and something like ~7.7ms with AMDVLK (or AMDGPU-PRO) on Linux. I am running into some unrelated performance issues on Windows so can't test that yet. Haven't had time to investigate this yet.
- On all 3 Linux drivers and 1 Windows driver, it looks like unlike RDNA2, the cost of dispatching "empty" mesh workgroups via `EmitMeshTasksEXT(0, 1, 1)` is very high, much higher than NVidia or RDNA2 (at least on radv from last year; I don't have a capable RDNA2 to compare directly now...). This is unfortunate; I'm not sure if this is a hardware issue (which seems more likely as all drivers have the same behavior) or if this can be worked around in the drivers or shaders somehow. This impacts two-phase cluster occlusion culling, which means I'd need to figure out how to rework it - probably that means not using the task shaders in the first place :(

**Comment by zeux** [comment]
> On all 3 Linux drivers and 1 Windows driver, it looks like unlike RDNA2, the cost of dispatching "empty" mesh workgroups via EmitMeshTasksEXT(0, 1, 1) is very high

I'm not sure now whether RDNA2 ever worked well here. Maybe I just had a weak GPU I was testing - I don't have the previous GPU on hand anymore. So let's just focus on RDNA3.

I've looked into this closer and the default ring buffer sizes just make task shaders useless for efficient culling on 7900 GRE. This is on both radv and Windows AMD driver; on radv I was able to change the ring buffer sizes and confirmed dramatic speedups.

Specifically, in radv, niagara on kitten.obj full screen 4K resolution, I get:

- Default settings when cluster occlusion culling is disabled, first render pass takes 5.5ms
- Bumping ring buffer size (`ac_task_info::num_entries`, requires Mesa patch) from 1024 to 32768 makes it 5.0ms

- Default settings when cluster occlusion culling is enabled, first render pass takes 5.2ms, second render pass takes 5.0ms
- Bumping ring buffer size to 32768 makes first render pass 1.6ms (!), second render pass 0.4ms (!!).

Obviously the culling rejection rates are very good which is why this is such a problem. But this makes task shaders useless for efficient culling on RDNA3. You should be always better off using a compute shader. The problem with large ring buffer size, in addition to this just not being the default configuration, is that the memory impact is very significant because of task payload - 32K num_entries costs 512 MB of ring buffer memory because every task is allocated 16 KB.

My understanding is that the "empty" draw commands are just a fact of life on AMD, based on the comment in `task_draw_ready_bit`, and presumably the ring buffer being managed externally by CP:

```
    * Task shaders MUST write the draw ready bit to the draw ring
    * before they finish. The firmware waits for the shader to write
    * this bit before it reads the mesh dispatch size to launch the
    * mesh shader workgroups.
    *
    * If the task shader doesn't write this bit, the HW hangs.
```

Unfortunately I will need to completely abandon task shaders for cluster culling to actually work great on AMD.

**Comment by zeux** [comment]
Also for completeness here's RGP trace from Windows driver; two large yellow bars are two render passes, cluster occlusion culling is enabled here. You can see that the second pass takes forever and occupancy is really bad; the first pass though *also* has poor occupancy on geometry waves, just not as bad as the second pass. Obviously I can't easily rebuild the Windows driver with a different ring buffer size :) I haven't checked if that even uses the same ring management scheme but I'd assume it would even if the ring sizes are configured differently.

![image](https://github.com/user-attachments/assets/f25259f5-4bb3-498c-b97d-75d151ce4c47)

**Comment by Venemo** [comment]
Hello Arseny,

Sorry for the late reply here.

> It looks like the performance is much more consistent between different meshlet configurations; this should make it much easier to tune which is good.

Yes, this is thanks to row exports and the new fast launch mode on RDNA3.

> Software triangle culling in mesh shader doesn't seem as relevant as it was on RDNA2. This is also great, because it reduces the amount of tradeoffs involved wrt occupancy/ALU et al.

AFAIK, RDNA3 has improved the fixed-function culling in the rasterizer, so shader based culling is much less relevant, although the impact may be still non-zero if your mesh shader creates a lot of primitives and accesses a lot of memory.

> On radv, it looks like the mesh shading path in the default configuration (64/64) is a little slower vs AMDVLK with the same code. I see something like ~7.9ms for the render pass (4K full screen, 7900 GRE) with radv, and something like ~7.7ms with AMDVLK (or AMDGPU-PRO) on Linux. I am running into some unrelated performance issues on Windows so can't test that yet. Haven't had time to investigate this yet.

Can you share RGP captures comparing the two drivers in this scenario?

> On all 3 Linux drivers and 1 Windows driver, it looks like unlike RDNA2, the cost of dispatching "empty" mesh workgroups via EmitMeshTasksEXT(0, 1, 1) is very high, much higher than NVidia or RDNA2 (at least on radv from last year; I don't have a capable RDNA2 to compare directly now...). This is unfortunate; I'm not sure if this is a hardware issue (which seems more likely as all drivers have the same behavior) or if this can be worked around in the drivers or shaders somehow.

We can maybe ask AMD's team to look into this. Do you have something that demonstrates this problem?

I vaguely recall there was a recommendation to set the X dimension to zero when any other is zero, which we implemented:

```
      /* When either Y or Z are 0, also set X to 0.
       * Not necessary, but speeds up the job of the CP.
       */
      x = nir_bcsel(b, nir_ieq_imm(b, nir_ior(b, y, z), 0), nir_imm_int(b, 0), x);
```

At a wild guess, maybe it would help to set all of them to zero?

> I've looked into this closer and the default ring buffer sizes just make task shaders useless for efficient culling on 7900 GRE. This is on both radv and Windows AMD driver; on radv I was able to change the ring buffer sizes and confirmed dramatic speedups.

Have you tried other buffer sizes between 1024 and 32768? I imagine that it would hit diminishing returns after some point. Alternatively, have you tried to increase the workgroup size of your task shader to process more meshlets in a single workgroup?

If there is significant gains to be had, we could make the task payload allocation more dynamic (ie. allocate less payload when the shaders don't use the full 16K) and therefore enable higher throughput.

> Unfortunately I will need to completely abandon task shaders for cluster culling to actually work great on AMD.

I think this isn't the first time I hear from this issue, but you are the first who actually shared concrete details. I would be happy to work with you to improve the driver to behave better with task shaders on our driver (RADV).

**Comment by zeux** [comment]
> Have you tried other buffer sizes between 1024 and 32768? I imagine that it would hit diminishing returns after some point.

Yeah intermediate buffer sizes are also helpful, just less so. I think 32K is the point at which I stop seeing gains, but would need to test more.

Increasing the work group size is not very helpful: it is 64 right now, and it can't be set larger than 128 due to min spec limits; 128 is slightly better vs 64 but still hits the same bottleneck (eg it's 5.2 ms => 5.0 ms).

I think it's really unfortunate that the payload is fixed at 16K right now; my payload is something like 260 bytes (4b plus 4b per thread which I have 64 of at the moment, so 4+64*4), so there's way more task entries available in the same memory space. If the payload selection was more dynamic and the task queue size adjusted accordingly, this code would automatically see dramatically better performance; neither radv nor AMDVLK seem to be doing this. But I'm not sure if this is a fundamental limitation (eg if ring buffers are too expensive to reconfigure?).

**Comment by zeux** [comment]
Using this small patch and running with different values of RADV_TASKN (and RADV_DEBUG=nocache to disable shader caching as the value is embedded):

```
diff --git a/src/amd/common/ac_gpu_info.c b/src/amd/common/ac_gpu_info.c
index 5b99ee54a8d..4fbec269382 100644
--- a/src/amd/common/ac_gpu_info.c
+++ b/src/amd/common/ac_gpu_info.c
@@ -2435,8 +2435,11 @@ void ac_get_hs_info(const struct radeon_info *info,
    hs->tess_offchip_ring_size = hs->max_offchip_buffers * hs->tess_offchip_block_dw_size * 4;
 }
 
-static uint16_t get_task_num_entries(enum radeon_family fam)
+static uint32_t get_task_num_entries(enum radeon_family fam)
 {
+   char* env = getenv("RADV_TASKN");
+   if (env) return atoi(env);
+
    /* Number of task shader ring entries. Needs to be a power of two.
     * Use a low number on smaller chips so we don't waste space,
     * but keep it high on bigger chips so it doesn't inhibit parallelism.
@@ -2463,7 +2466,7 @@ static uint16_t get_task_num_entries(enum radeon_family fam)
 void ac_get_task_info(const struct radeon_info *info,
                       struct ac_task_info *task_info)
 {
-   const uint16_t num_entries = get_task_num_entries(info->family);
+   const uint32_t num_entries = get_task_num_entries(info->family);
    const uint32_t draw_ring_bytes = num_entries * AC_TASK_DRAW_ENTRY_BYTES;
    const uint32_t payload_ring_bytes = num_entries * AC_TASK_PAYLOAD_ENTRY_BYTES;
 
diff --git a/src/amd/common/ac_gpu_info.h b/src/amd/common/ac_gpu_info.h
index 954b7a3a2ae..d790614b145 100644
--- a/src/amd/common/ac_gpu_info.h
+++ b/src/amd/common/ac_gpu_info.h
@@ -358,7 +358,7 @@ struct ac_task_info {
    uint32_t draw_ring_offset;
    uint32_t payload_ring_offset;
    uint32_t bo_size_bytes;
-   uint16_t num_entries;
+   uint32_t num_entries;
 };
 
 /* Size of each payload entry in the task payload ring.
```

With the methodology of:
- change niagara src/config.h to disable vsync, build in Release to disable validation
- run ./niagara data/kitten.obj
- keep window at current size (note, the numbers are a little different from above because of this but same ballpark)
- note the first render time in title bar
- press K to enable cluster occlusion culling
- note the first and second render times in title bar

I get this on 7900 GRE:

TASKN | render (no COC) | render early (COC) | render late (COC)
--|--|--|--
1024 | 5.34 | 5.0 | 4.8
2048 | 5.35 | 4.9 | 4.7
4096 | 5.35 | 4.7 | 4.6
8192 | 5.35 | 4.2 | 4.0
16384 | 5.25 | 3.0 | 2.5
32768 | 4.98 | 1.6 | 0.3
65536 | 4.98 | 1.6 | 0.3
131072 | 4.98 | 1.6 | 0.3

Note that the plateau happens at 32K but I just changed the sizes without checking the consequences; if there is somewhere else where the ring buffer index is limited to 16 bits then maybe the plateau is further along.

**Comment by Venemo** [comment]
> Increasing the work group size is not very helpful: it is 64 right now, and it can't be set larger than 128 due to min spec limits; 128 is slightly better vs 64 but still hits the same bottleneck (eg it's 5.2 ms => 5.0 ms).

RADV exposes 1024 for that, so it should work. I agree that this would need to be configurable on your side, since other drivers (NVidia?) wouldn't allow that much. It would definitely be worth an experiment though, because larger task shader workgroups would require less entries in the ring buffer.

> I think it's really unfortunate that the payload is fixed at 16K right now; [... ] neither radv nor AMDVLK seem to be doing this. But I'm not sure if this is a fundamental limitation (eg if ring buffers are too expensive to reconfigure?).

In fact I think the vast majority of users don't use too much payload; every mesh shading best practices talk I've seen suggests to use as little payload as possible.

The truth is that when we were working on the driver, there was barely any apps to test with, so we were building the technology without any users, which lead to some usecases left unconsidered (or less performant). It's not impossible to reconfigure the ring buffers, but maybe we never had a test case when the task shaders cull that much.

**Comment by zeux** [comment]
Oh and to give an idea of what the culling rates are:

- "render (no COC)" pass would output approximately 80% of meshlets from task shader culling stage, so ~51 mesh workgroups from each task workgroup
- "render early (COC)" pass has much better culling rates and would output approximately 25% of meshlets from task shader culling stage, so ~16 mesh workgroups from each task workgroup
- "render late (COC)" pass in this scenario outputs *no* mesh workgroups, as it is confirming the occlusion false positive rates for correctness, so all dispatches it outputs are empty.

**Comment by zeux** [comment]
> RADV exposes 1024 for that, so it should work.

The issue is that glslang enforces 128 as a default. I assume it can maybe be pushed further with that limit overridden.
The other problem is that a value that is too large may stop being useful as I do not group different meshes in the same task workgroup; with 64 meshlets * 96 triangles it's already 6K triangles which is larger than my smallest LOD iirc so I'd expect some task groups maybe have fewer; with 256 you'd be looking at very poor utilization.

**Comment by Venemo** [comment]
We are just discussing with Marek what to do about this.

Making the buffer entirely dynamic is very problematic because we'd have to not only reconfigure it, but also wait for previous draws (with a different buffer configuration) to finish.

A simpler solution would be to allocate 2 separate buffers, one that would be used when the payload is small (maybe 0.5K or 1K per entry?) and another when the payload is large (16K per entry). Then, we could reconfigure in runtime which buffer to use, without any loss of runtime performance.

This still raises some questions regarding what happens when the payload is mismatched between task and mesh, but I assume that reading uninitialized payload memory is UB, so it would be enough to just care about the payload usage of the task shader.

Would you be willing to write a microbenchmark? A small VK app to expose the problematic cases with `EmitMeshTasksEXT(0, ...)` and also another to expose the problem with low occupancy with high task cull rates. I think that would help both us and AMD to look into fixing these problems.

**Comment by zeux** [comment]
This repository should be useful as a benchmark I think. It would be better if it had a mode where it gathered results in a headless mode without having to open a window and inspect numbers changing in realtime. I can add that.

And yeah ack on having to synchronize between draws of different configurations. I could see this being a problem in larger apps.

**Comment by zeux** [comment]
> also wait for previous draws (with a different buffer configuration) to finish

I'm curious, how is this done right now? If a buffer is reused with the same configuration you still get slots from the previous draw being read by the in-flight mesh workgroups, and they are written by the subsequent task shader. I don't see code in ac_nir_lower_taskmesh_io_to_mem that synchronizes accesses, it seems to just write payload and draws directly.

**Comment by Venemo** [comment]
> I'm curious, how is this done right now?

We emit the `DISPATCH_TASK_STATE_INIT` packet in the preamble (the preamble contains some commands that are executed before the contents of the submitted command buffers) as part of `radv_emit_task_rings`. That tells the firmware where the "control buffer" is. The control buffer is initialized in `radv_initialise_task_control_buffer` and lets the firmware know the address of the draw ring buffer and the number of entries in it.

When launching a task-mesh draw, we have to submit a command buffer to a compute queue (this is a HW detail that isn't visible to the application and it will execute task shaders only) and another on a graphics queue (this will execute mesh shaders and everything else that the application wants). The firmware is able to launch as many task shader workgroups as available entry in the draw ring buffer, so each task workgroup gets a draw ring entry. The firmware keeps track of which entries are used and waits until the task shader writes the "draw ready bit"; then it launches mesh shader workgroups according to the dispatch dimensions that were written in the draw ring. [This XDC talk](https://www.youtube.com/watch?v=9sJO4Oxb0F8) has more details if you're interested. Once the mesh shader workgroups are finished, the same entry can be reused by a different task shader and the mesh workgroups launched by that.

As far as I understand the process, multiple task-mesh draws can be in flight at the same time and can share the same draw ring buffer with the above mechanism; although I am not fully confident in how that is handled exactly.

The payload ring is entirely managed by the driver: the shader basically addresses it using the same ring entry index that it uses for the draw ring (so there is a requirement of having the draw ring and payload ring have the same amount of entries).

I hope that answers your question, but let me know if you are curious about more details.

**Comment by Venemo** [comment]
Not sure if my above wall of text explains it clearly, but currently we never reconfigure the ring buffer; we always have the same amount of entries and allocate the same size; so we never had to think about what happens in the firmware when you set a different or the same control buffer with different number of entries. Either the firmware is smart enough to have multiple task-mesh draws with different control buffers in flight; or `DISPATCH_TASK_STATE_INIT` causes an implicit wait.... or we would need a manual wait.

**Comment by Venemo** [comment]
> The reason I asked is the scheme you suggested would certainly work but it's a little uncomfortable to have such a steep cliff for crossing a single payload threshold.

We could do it more dynamically, but that's problematic because the draw ring entries are always the same size. So we would need to allocate a large enough buffer to hold more draw ring entries; and then reconfigure the number of entries based on the available space in the payload ring. This would also mean moving out the configuration from the preamble into the command buffers, which is annoying.

What do you think, what is a typical use? How much payload does a typical application use? Regardless of what implementation we end up with, I'd like to choose a sensible default so that reconfiguration doesn't have to occur much in the real world.

**Comment by zeux** [comment]
It's hard to answer this because I don't think there's much task shader usage in the wild, if any. Several engines that use mesh shading in production do *not* use task shaders; my understanding is that that's motivated by the same issue we are exploring here. The couple open source engines/apps I know of that do are squarely under the 512 byte limit.

NVidia demos actually tend to use even smaller payloads; my code in this repository uses ~4 bytes per thread (so ~256b for 64-thread workgroup), but their examples actually use ~1. Another open source engine I found that uses amplification shaders in DX uses ~2 bytes per thread because of some HLSL limitations. My code only uses 4 bytes because this didn't come up before as a bottleneck, I can easily shrink this down to ~1 byte per thread (so even larger work groups will have <512 payload).

From this perspective I think it's reasonable to have the "fast path" configuration be 512 bytes, certainly not larger, because I don't actually know an example OTOH that uses that.

**Comment by zeux** [comment]
Two questions I missed:

> Can you share RGP captures comparing the two drivers in this scenario?

I don't have RGP set up on Linux atm unfortunately, the RGP capture above was from Windows. I'd also need to remeasure vs latest Mesa, maybe this discrepancy has gone away.

> I vaguely recall there was a recommendation to set the X dimension to zero when any other is zero, which we implemented [...] At a wild guess, maybe it would help to set all of them to zero?

That doesn't seem to change performance (as in, 0,1,1 seems equivalent to 0,0,0 in my tests).

**Comment by Venemo** [comment]
> I don't have RGP set up on Linux atm unfortunately, the RGP capture above was from Windows. I'd also need to remeasure vs latest Mesa, maybe this discrepancy has gone away.

You don't really need a special setup to take a RGP capture with RADV. (In fact RGP isn't integrated with RADV.) You can take a capture from RADV and open that with RGP on your Windows setup.

**Comment by zeux** [comment]
Thanks, this works. I captured some profiles but they are somewhat difficult to compare exactly between Windows and Linux (radv); unsure if there are clock differences to account for, but also I don't have the AMDVLK/Linux working anymore since I upgraded to Ubuntu 24.10 last week and there is no compatible amdgpu package for that yet (and I'm not building it from source :D). I'll come back to this discrepancy (if it still exists) when AMD updates their official drivers for Linux.

**Comment by Venemo** [comment]
> Thanks, this works. I captured some profiles but they are somewhat difficult to compare exactly between Windows and Linux (radv)

Does the discrepancy exist between the Windows driver and RADV? If it does, it should be visible in RGP

**Comment by zeux** [comment]
I'm not sure - it's complicated. The discrepancy exists (between Windows & Linux/radv) according to the overall frame rate and the individual GPU timers as captured in real time. This is with vsync off so in theory I'd expect similar clocks when just running the app (there's little idle time), but I would need to verify this. When running radv on Windows, I see nominal performance before I capture to be the same - higher than radv - but when I capture, the resulting capture has longer duration that matches what radv captures more or less. radv captures also show longer capture duration but it's less severe (eg radv is 5ms frame time with no capture, 5.2ms in RGP capture; Windows is 4.5ms frame time with no capture, 5.2ms in RGP capture).

Additionally, when using cluster culling with the aforementioned second pass experiencing long delays due to task queue bottleneck, the RGP profiles just look very different between radv and Windows driver - the total number of vkCmdDispatch that radv shows is a close multiple of 1024 (queue size), whereas I am dispatching fewer mesh workgroups; Windows RGP capture shows a more reasonable size with a much shorter duration but there's still a gap at the end where "nothing" happens on the timeline, so I'm not sure if this is a capture artifact or not.

Consequently there's a lot of variables here, and to test any theory on both OSes I have to reboot back and forth which is time consuming, so it would be better to compare radv vs amdvlk, but for that I need an amdvlk binary that just works so I'll wait :)

**Comment by Venemo** [comment]
Understood. Regardless of what RGP says, it is nice to see that the two drivers at least perform in the same ballpark.

**Comment by Venemo** [comment]
FWIW, the number of draw entries is 1024, so theoretically 1024 task shader workgroups could be in flight on the GPU at any given time.

Considering that the 7900 GRE has 80 CUs (note, the top Navi 31 has up to 96 CU) and each CU has 2 SIMD, and each SIMD can have 16 waves in flight, that means 7900 GRE can have 80*2*16=2560 waves in flight at a time (this is 96*2*16=3072 on top Navi 31) . Now the task shaders here are just 1 Wave64 wave per workgroup, meaning we should be able to have at least 40% occupancy on your GPU with just task shaders (this would be 33% on top Navi 31), assuming each task workgroup launches 0 mesh shaders, but in reality the occupancy seems to be much lower. Theoretically it should be enough to increase the number of entries to 4096 (closest power of two after 3072) to get fully occupancy, but that doesn't seem to help either.

So, I suspect that either the issue is due to the firmware; or the driver does something really stupid that nobody has noticed.

**Comment by zeux** [comment]
I agree that this doesn't fully make sense. I originally expected that this is just an issue with CP throughput, but thinking about this further, unless CP has some sort of parallel draw call rejection feature, increasing the queue size would not by itself allow the "empty" dispatch pass to accelerate more than 2x (by allowing CP & task shaders to run in parallel). And if CP is not the bottleneck then it's unclear why the queue needs to be increased so much to have good gains, and why modest increases in queue sizes don't seem to help much. Also I was discussing this with a dev who had a DX12 playground with very different code, and they observed similar timings for empty dispatches with a lighter weight task shader, so that also seems to suggest CP bottleneck, which again is a little odd as to why the gains are so dramatic.

But I could imagine some sort of horrible synchronization ping ponging where both CP and TS are constantly stalled, with CP waiting for some items to be written ahead of current read pointer and TS waiting for available space in the ring buffer, that magnifies the slowdown well past 2x and requires read and write regions of the ring buffer to be far apart to isolate them from conflict.

It's also possible this is some sort of driver issue, but because the synchronization is entirely on the firmware side and because this is present in all drivers (radv, amdvlk/linux, amdvlk/windows, although the latter two are the same source code of course), I'd think this is either some inefficiency in firmware, or a fundamental design defect of the hardware. FWIW I've hit the same type of problem - bottleneck by empty draws in CP processing - on RDNA2 on this code base, when I was using multi draw indirect and had multiple draw calls; this was fixed for RDNA2 by switching to a single draw call (https://github.com/zeux/niagara/commit/463063ccc9d4ca83f4ade29cb91541672882558e commit, https://www.youtube.com/live/eYvGruGHhUE YT stream); I've went back to this commit which allowed to switch between two modes, and this is still super valuable on RDNA3, but wheres on the GPU I had at the time (6700 XTX) doing this alone basically lifted the CP bottleneck, on 7900 GRE this is very beneficial but not enough to eliminate the CP bottleneck. I'm not sure if this is because the firmware was different, or because just the delta between shader processing capacity and CP processing capacity or something else changed.

**Comment by zeux** [comment]
Oh one other thing that complicates correct comparison here that I think I should note: enabling rgp trace in the 32K queue size configuration *slows things back down*.

Specifically, when I use cluster culling with the slow second pass with default 1024 queue size, I get the following timings (sorry that these are different from before, I've been optimizing some unrelated bits and pieces, treat these in isolation from previous timings):

gpu render early 4.6ms
gpu render late 4.5ms
gpu render frame 9.34 ms
cpu observed render latency 10.0 ms

When I use queue size 32k, I get the following: 

gpu render early 1.2ms
gpu render late 0.4ms
gpu render frame 1.75ms
cpu observed render latency 2.0 ms

I am running these in a CPU-GPU synced configuration (CPU waits for GPU frame completion) so I am inclined to trust the results because CPU timing can not be spoofed even if GPU counters somehow measure the wrong thing.

However, when I enable MESA_VK_TRACE=rgp, I get this with 32K entries (the numbers with 1024 entries are same ballpark as without rgp):

gpu render early 5.2ms
gpu render late 5.2ms
gpu render frame 10.7ms
cpu observed render latency 11.4ms

This makes it especially difficult to correctly measure all of this, which is part of the "it's complicated" I referred to: I do not trust RGP captures to not skew numbers to the point of not being useful in this code...

The RGP capture for 32K queue does show a huge gap in the second pass where seemingly nothing happens. Maybe this is the actual capture done by firmware (?).

![image](https://github.com/user-attachments/assets/f13f4fc8-9982-4410-bafb-78b3def94d02)

**Comment by Venemo** [comment]
I think it makes sense what you say, that the bottleneck is the CP firmware, though I personally have no insight into what is happening in there exactly. I wouldn't be surprised, since it seems that both of these GPUs are all bottlenecked by the CP in general, and the 7900 even more so.

Based on what you posted, it looks like taking a RGP trace will by itself slow down everything to such an extent that we are essentially looking at a different thing entirely. If you have time, you could try setting `THREAD_TRACE_MARKER_ENABLE` to zero in `radv_cs_emit_dispatch_taskmesh_gfx_packet` and see what happens then. (My guess is that we'll see less information in RGP but maybe the slowdown will change.)

Sorry I haven't been able to try it myself yet, I just recently got back from XDC 2024. I would like to test this on my 7900 XTX too. Can you please give me a quick walk through of what I need to do to run your test cases?

**Comment by zeux** [comment]
> setting THREAD_TRACE_MARKER_ENABLE to zero in radv_cs_emit_dispatch_taskmesh_gfx_packet

Yeah this basically eliminates most of the tracing overhead AFAICT; this is with that market set to 0 and with queue items 1024:

![image](https://github.com/user-attachments/assets/e1fc7539-6420-4465-9cac-85f5a1ae3f13)

... and this is with queue items 32768 (both captures are with cluster culling enabled):

![image](https://github.com/user-attachments/assets/8475332d-cf8c-4b0c-bac9-9745c96fc8b8)

The overall frame times barely change under RGP in that setup. It would be nice to have this as a tracing option maybe, so that it could be disabled for testing draw heavy workloads...

> Sorry I haven't been able to try it myself yet

No worries at all, this is extremely not urgent :)

If you'd like to reproduce any of the results yourself, it should be sufficient to:

```sh
git clone --recursive https://github.com/zeux/niagara
cd niagara
cmake . -DCMAKE_BUILD_TYPE=Release
make -j8
./niagara data/kitten.obj
```

By default, the rendering pipeline uses mesh shaders, with task shaders that cull clusters based on frustum/backface, but *do not* do per-cluster occlusion culling, which is what causes the second pass to have to reprocess all meshes but output zero mesh shading workgroups. You can toggle cluster occlusion culling by pressing K. The other useful key is M, which disables mesh/task pipeline altogether (as well as any forms of cluster culling!) and just uses traditional raster. Because the CP bottleneck affects the first pass as well, without a large ring buffer capacity that is much slower than mesh shading path, at least on 7900 GRE; with patched driver and 32768 ring size, I get similar timings.

The title bar shows the total CPU latency, GPU frame time (measured with GPU timestamps), individual frame times for first and second render passes, and some other information that can probably be ignored. Note that the window size will affect the timings - while this is a geometry-heavy scene, culling and LOD selection take window size into account.

For testing ring sizes, I use the patch mentioned in https://github.com/zeux/niagara/issues/30#issuecomment-2407688500 on top of latest (branch main) radv, built with LLVM disabled.

By default, VSYNC is *enabled*; to disable it, just change the CONFIG_VSYNC variable in `src/config.h` and run `make` again. This should not affect GPU timings, unless the GPU power management decides to keep GPU clocks fairly low. When available, my code uses KHR_performance_query to activate profiling lock; this used to work on 6700 XTX and fixed the GPU clocks, but it looks like radv (or amdvlk) no longer exposes it on 7900 so the app is at the mercy of the default power management logic.

**Comment by zeux** [comment]
One more comment on RDNA2: while I no longer have access to 6700 XTX, I have an integrated RDNA2 in my Zen4 CPU. It's obviously a GPU of a completely different class, but what I found interesting is that when testing using the same patch, the CP bottleneck doesn't appear to exist (the frame times are pretty flat with ring size varying all the way from 256 to 16K), however using a queue size of 32K sees a significant jump *up* in render time for both passes, in the configuration where cluster culling is enabled. I'm not sure why this is exactly; perhaps CP needs to process the entire buffer's worth of dispatches generated, and that results in excessive processing by CP for draw commands that aren't generated; but this points to either RDNA2 being just different, or to much weaker GPU potentially wanting smaller queue sizes (without an integrated RDNA3 to test I can't disambiguate the two).

**Comment by Venemo** [comment]
> I have an integrated RDNA2 in my Zen4 CPU. [...] using a queue size of 32K sees a significant jump up in render time for both passes

I think it's not a surprise that a small GPU would have different performance characteristics. My guess is that the CP in those GPUs is roughly the same, but it has (much) fewer compute units to feed. Furthermore, memory access is much slower on APUs than on dGPUs.

Considering that Navi 31 can have max 96 x 2 x 16 = 3072 waves in flight, and a 32K ring bufer works best, I would extrapolate that on a GPU that has only 2 CUs, since it can have max 2 x 2 x 16 = 64 waves in flight, the ring buffer could be as small as 512 or 1024. Technically, since Raphael probably has other bottlenecks, it doesn't matter as much os on a dGPU.

Also, I just noticed that all of the calculations I posted here are rendered incorrectly due to `*` being understood as italic text... my apologies for that.

**Comment by zeux** [comment]
> I think it's not a surprise that a small GPU would have different performance characteristics.

Yes, I agree with that. My point was that the strategy of maximizing the queue size for small payloads backfires on this GPU for some reason. Maybe some other adjustments somewhere else are needed, or the queue size could be scaled with the CU count.

**Comment by Venemo** [comment]
> My point was that the strategy of maximizing the queue size for small payloads backfires on this GPU for some reason.

Have you tried 512 or 1024 on that small GPU? That would be equivalent to 32K on the large GPU.

**Comment by zeux** [comment]
As noted above, all sizes from 256 up to 16K perform the same. Only 32K is an outlier (and results in significant, 50%+, frame time regression).

**Comment by zeux** [comment]
Final note on this: I've looked into the discrepancies in performance I've seen here with AMDVLK. Using latest Mesa build (as of today, 6800cd270306ad779b72bbed754bbcf463d1c78c) + https://github.com/GPUOpen-Drivers/AMDVLK/releases/tag/v-2024.Q3.3 AMDVLK & latest master in this repository (as of today, 4242461f84532c3c4b14b6f7f29d59be05e09217).

All numbers are in 4K resolution using Wayland with VSYNC off. This is a larger resolution vs what I was using before, and other updates in the repository shifted the numbers; as before, numbers in this comment should only be understood in isolation / relative to each other, not relative to previous reports.

Latest master now defaults to *not* using task shaders (which is a new mode I've implemented on a stream a week-ish ago, https://youtu.be/zROUBE5pLuI). This mode eliminates the overhead associated with task dispatch by using a single mesh grid with a compute shader replacing task shader (running in graphics queue and generating cluster ids). That's significantly faster, has no CP overhead, and allows to both test mesh shader performance in isolation and *actually* stress the parts of the GPU pipeline that are downstream of CP.

In that mode, I see radv and amdvlk at ~parity with or without cluster occlusion culling:

radv: 2.20 ms/frame with cluster occlusion, 4.30 ms/frame without it.
amdvlk: 2.23 ms/frame with cluster occlusion, 4.26 ms/frame without it.

With task shading and mesh shading disabled (traditional raster), I see radv being slightly faster:

radv: 4.81 ms/frame
amdvlk: 4.93 ms/frame

Now, *with* task shading and mesh shading enabled (which is the configuration that was the default in my previous comment about task shading performance; now it's optional and is enabled by pressing `T`), I see the performance regressions that I've commented on previously:

radv: 13.28 ms/frame with cluster occlusion, 7.16 ms/frame without it.
amdvlk: 11.78 ms/frame with cluster occlusion, 7.06 ms/frame without it.

The times are much larger than other reported times because of the CP bottleneck we've already discussed. And, as already explained, cluster occlusion culling in this configuration is actually detrimental to overall performance because CP bottleneck makes the second pass too slow to be usable.

As I noted in the previous comment, AMDVLK actually scales the ring buffer with the GPU configuration, and my GPU should be getting 8192 items in the ring buffer, instead of radv's default 1024. Changing radv to use 8192 ring size changes the results to:

RADV_TASKN=8192 radv: 11.67 ms/frame with cluster occlusion, 6.98 ms/frame without it.

So the previous performance delta I've observed is entirely explained by the ring size.
Using larger rings, as before, makes things better; as before, 32K is a significant jump over 16K:

RADV_TASKN=16384 radv: 9.31 ms/frame with cluster occlusion, 6.58 ms/frame without it.
RADV_TASKN=32768 radv: 3.77 ms/frame with cluster occlusion, 5.38 ms/frame without it.

So! The good news is that radv is consistently outperforming amdvlk; the only case when that's not happening is the one where radv is configuring the task queue without scaling it according to the GPU shader engines, which yields a lower queue size vs amdvlk and consequently worse performance.

The bad news is that not using task shaders is significantly faster than even the configuration with 32K items.

I'm going to close this issue because I don't think I can meaningfully provide more input here. My conclusions are:

1. Both AMD drivers *really* need to at least scale the ring up according to the payload count. The current performance cliff is too large for task shaders to be practically useful for culling. This conclusion is shared by games like Alan Wake2 that do cluster culling in compute instead of using amplification shaders per Digital Dragons talk (https://www.youtube.com/watch?v=EtX7WnFhxtQ). Doing this optimally seems to require detailed information about the way the CP handles item readiness. I hope this happens for both drivers; until it does, task shaders have limited utility for geometry heavy scenes on AMD.

2. Even if the ring is scaled up to 32K, the performance is still not competitive with a manual replacement (in comparison, on NVidia 4090 on this application, task shader culling is the same performance as manual compute culling, while requiring less space since the entire cluster buffer doesn't need to be stored in memory). It is possible that on real-world scenes, the discrepancy between manual culling and task culling will not be visible and the data I am seeing is an artifact of synthetic setup. Until pt1 is resolved that will be very hard to tell; but I also hope AMD is considering fixing this in firmware or, if necessary, in hardware in future revisions of whatever is the issue here; without this, task shaders work very well on NVidia and poorly on AMD which is not a great place to be.

**Comment by zeux** [comment]
Yes, you can still activate task shader mode by pressing `T`. The comment above details the measurements in that mode from master branch.
